Panos network operations hygiene base

adambaumeister opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
This is the next series of pan-os commands for the integration using the pan-os-python SDK.

This PR includes two commands that are very difficult to write into test playbooks or command examples, as they can break lab enviornments - `pan-os-platform-install-software` and `pan-os-platform-reboot`. Neither returns much data or implements any logic outside of what is covered by unit testing, so I've omitted them from the test playbooks and command examples for safety reasons.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [X] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [X] Tests
- [X] Documentation

GuyAfik reviewed (commented):
Hi @adambaumeister, nice work as usual!

like previous PRs here's a few points:
1) make sure there aren't any secrets in the README.md file.
2) lets go over together on which commands should be in the test-playbooks and which shouldn't.
3) make sure in unit-tests you cover all the new commands.
4) on commands like reboot - maybe its worth to add to the documentation what are the Consequences when running it (if exists).

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7706,6 +7707,18 @@ def all(self, filter_string: Optional[str] = None) -> Iterator[Union[Firewall, P
         for device in all_devices.values():
             yield device
 
+    def one(self, filter_string: str) -> Union[Firewall, Panorama]:
```
good call.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7706,6 +7707,18 @@ def all(self, filter_string: Optional[str] = None) -> Iterator[Union[Firewall, P
+        """
+        all_devices = {**self.firewall_objects, **self.panorama_objects}
+        if filter_string not in all_devices:
+            raise DemistoException(f"filter_str {filter_string} does not exactly match any devices known to this topology.")
```
Good idea! Added.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8601,6 +8614,314 @@ def resolve_container_name(container: Union[Panorama, Firewall, DeviceGroup, Tem
+    def __init__(self, register: dict):
+        self.register: dict = register
+
+    def get(self, issue_code) -> ConfigurationHygieneCheck:
```
Fixed this - it's a string. Also fixed the type annotation for the register attribute.

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8601,6 +8614,314 @@ def resolve_container_name(container: Union[Panorama, Firewall, DeviceGroup, Tem
+            device_filter_str: str = None,
+    ) -> list[PanosObjectReference]:
+
+        result = []
```
This is good but i ended up removing that function as it was unneeded

adambaumeister commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8601,6 +8614,314 @@ def resolve_container_name(container: Union[Panorama, Firewall, DeviceGroup, Tem
+                        issue_code="BP-V-2"
+                    ))
+                    check = check_register.get("BP-V-2")
+                    check.result = UNICODE_FAIL
+                    check.issue_count += 1
+
+                required_log_types = ["traffic", "threat"]
+                for log_forwarding_profile_match_list in match_list_list:
+                    if log_forwarding_profile_match_list.log_type in required_log_types:
+                        required_log_types.remove(log_forwarding_profile_match_list.log_type)
+
+                for missing_required_log_type in required_log_types:
+                    issues.append(ConfigurationHygieneIssue(
+                        hostid=resolve_host_id(device),
+                        container_name=resolve_container_name(container),
+                        description=f"Log forwarding profile missing log type '{missing_required_log_type}'.",
+                        name=log_forwarding_profile.name,
+                        issue_code="BP-V-2"
+                    ))
+                    check = check_register.get("BP-V-2")
+                    check.result = UNICODE_FAIL
+                    check.issue_count += 1
+
+        # No logging profiles configured in environment - BP-V-1
+        if len(lf_profile_list) == 0:
+            issues.append(ConfigurationHygieneIssue(
+                hostid="PLATFORM",
+                container_name="",
+                description="No log profiles configured!",
+                name="",
+                issue_code="BP-V-1"
+            ))
+            check = check_register.get("BP-V-1")
+            check.result = UNICODE_FAIL
+            check.issue_count += 1
+
+        return ConfigurationHygieneCheckResult(
+            summary_data=[item for item in check_register.values()],
+            result_data=issues
+        )
```
I think lets keep it the same, you're right it is fairly large but the inner loop is just iterating over the same set of objects anyway. I'll try to split newer functions up a bit further.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -7716,6 +7717,19 @@ def all(self, filter_string: Optional[str] = None) -> Iterator[Union[Firewall, P
+        :param filter_string: The exact ID of the device to return from the topology.
+        """
+        all_devices = {**self.firewall_objects, **self.panorama_objects}
+        if filter_string not in all_devices:
+            raise DemistoException(f"filter_str {filter_string} is not the exact ID of a host in this topology; " +
+                                   f"use a more specific filter string.")
+
+        return all_devices.get(filter_string)
```
always better to go on positive rather than negative 

```suggestion
        if device := all_devices.get(filter_string):
           return device
        raise DemistoException(f"filter_str {filter_string} is not the exact ID of a host in this topology; " +
                                   f"use a more specific filter string.")

```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8611,6 +8625,319 @@ def resolve_container_name(container: Union[Panorama, Firewall, DeviceGroup, Tem
+        Gets a single Hygiene check by it's string issue code.
+        :param issue_code: The string issue code, such as BP-V-1
+        """
+        issue_check = self.register.get(issue_code)
+        if not issue_check:
+            raise DemistoException("Invalid Hygiene check issue name")
+        return issue_check
```
```suggestion
        if issue_check := self.register.get(issue_code)
            return issue_check
        raise DemistoException("Invalid Hygiene check issue name")
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8611,6 +8625,319 @@ def resolve_container_name(container: Union[Panorama, Firewall, DeviceGroup, Tem
+                for rule_severity in rule.severity:
+
+                    # If the block severities are blocked
+                    if any(block_actions) and rule_severity in block_severities:
+                        block_severities.remove(rule_severity)
+                        if rule_severity in alert_severities:
+                            alert_severities.remove(rule_severity)
+                    # If the alert severities are blocked
+                    elif any(block_actions) and rule_severity in alert_severities:
+                        if rule_severity in alert_severities:
+                            alert_severities.remove(rule_severity)
+                    # If the alert severities are alert/default
+                    elif any(alert_actions) and rule_severity in alert_severities:
+                        if rule_severity in alert_severities:
+                            alert_severities.remove(rule_severity)
```
extract ```any(block_actions)``` and ```any(alert_actions)``` into variables and define them before the for loop, no reason for the any function to run over and over again in every iteration.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -9327,6 +9660,86 @@ def download_software(topology: Topology, version: str,
+    :param version: software version to upgrade to, ex. 9.1.2
+    :param sync: If provided, runs the download synchronously - make sure 'execution-timeout' is increased.
+    """
+    _sync = argToBoolean(sync)
+    return UniversalCommand.install_software(topology, version, device_filter_str=device_filter_string, sync=_sync)
```
```suggestion
    return UniversalCommand.install_software(topology, version, device_filter_str=device_filter_string, sync=argToBoolean(sync))
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -7466,6 +7466,195 @@ script:
+      description: New HA State
+      type: String
+  - arguments:
+    - default: false
+      defaultValue: null
```
what's the difference between these two?

adambaumeister replied:
I removed the `defaultvalue: Null` values as they do nothing. However, my understanding is as follows;

`default: true` flags that the argument has a default value (`true`) implemented in code 
`defaultValue: True` passes the value (`True`) into the argument by default unless it is overridden.

adambaumeister replied:
further, i think the reason you would use the default flag over defaultValue is if the argument is inherited from integration settings but overrideable, whereas you would use defaultValue if the argument simply has a default you want the user to see in the command.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/README.md:
```diff
@@ -5772,3 +5772,279 @@ Download the provided software version onto the device.
+| hostid | ID of host (serial or hostname) to reboot. | Required | 
+
+
+#### Context Output
+
+| **Path** | **Type** | **Description** |
+| --- | --- | --- |
+| PANOS.RestartStatus.Summary.hostid | String | Host ID | 
+| PANOS.RestartStatus.Summary.started | String | Whether download process has started. | 
```
can you add context example? make sure you add it for all the rest of the commands (you have already added it to some of the commands). 

Also add human readable section if possible.

adambaumeister replied:
Yeah, the ones I didn't add it to I couldn't build command examples for. I added context output for these but no human readable because it's a pain to write the markdown manually. Can add if absolutely required.

GuyAfik replied:
if its not too much of effort, it could be nice to have a simple example.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8601,6 +8614,314 @@ def resolve_container_name(container: Union[Panorama, Firewall, DeviceGroup, Tem
+        if not minimum_alert_severities:
+            minimum_alert_severities = BestPractices.VULNERABILITY_ALERT_THRESHOLD
+
+        conforming_profiles: Union[list[VulnerabilityProfile], list[AntiSpywareProfile]] = []
```
hasn't been done

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -9327,6 +9660,86 @@ def download_software(topology: Topology, version: str,
+    return HygieneLookups.check_vulnerability_profiles(
+        topology,
+        device_filter_str=device_filter_string,
+        minimum_block_severities=minimum_block_severities.split(","),
+        minimum_alert_severities=minimum_alert_severities.split(",")
```
```suggestion
        minimum_block_severities=argToList(minimum_block_severities),
        minimum_alert_severities=argToList(minimum_alert_severities)
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -7466,6 +7466,195 @@ script:
+      defaultValue: critical,high
+      description: csv list of severities that must be in drop/reset/block-ip mode.
+      isArray: false
+      name: minimum_block_severities
```
does it have specific values that it can have? if yes use a predefined section

adambaumeister replied:
It does not, there's only a set number of alert levels but the command only takes them in csv format (because you need to be able to specify multiple at once). As far as I am aware you can't use a list as a predefined argument so this is the only way we can do it.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -7466,6 +7466,195 @@ script:
+      description: csv list of severities that must be in alert/default or higher
+        mode.
+      isArray: false
+      name: minimum_alert_severities
```
does it have specific values that it can have? if yes use a predefined section

adambaumeister replied:
See above

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -9327,6 +9660,86 @@ def download_software(topology: Topology, version: str,
+    return HygieneLookups.check_log_forwarding_profiles(topology, device_filter_str=device_filter_string)
+
+
+def check_vulnerability_profiles(
+        topology: Topology,
+        device_filter_string: Optional[str] = None,
+        minimum_block_severities: str = "critical,high",
+        minimum_alert_severities: str = "medium,low"
+) -> ConfigurationHygieneCheckResult:
```
can you state in docstring whether this function should be used for both panorama and firewall or only for one of them?

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -7466,6 +7466,195 @@ script:
+      type: String
+    - contextPath: PANOS.ConfigurationHygiene.Summary.issue_count
+      description: Total number of matching issues
+      type: Nunber
```
```suggestion
      type: Number
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -7466,6 +7466,195 @@ script:
+      description: The shorthand code for this hygiene check
+      type: String
+    - contextPath: PANOS.ConfigurationHygiene.Summary.result
+      description: Whether the check passed or failed
+      type: String
```
are you sure its not a boolean here?

adambaumeister replied:
This one is actually a unicode string which renders into  a nice little ✔️  in the readable output on XSOAR :)

GuyAfik replied:
nice

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama_test.py:
```diff
@@ -1509,6 +1576,52 @@ def test_download_software(self, mock_topology):
+        # We also want to check that if an empty string is passed, an error is returned
+        with pytest.raises(
+                DemistoException,
+                match="filter_str  is not the exact ID of a host in this topology; use a more specific filter string."
```
the space here is in purpose?

adambaumeister replied:
Yes, this is the error message you get if you somehow pass an empty string into the filter string.

ShirleyDenkberg commented:
@GuyAfik Doc review completed.

GuyAfik reviewed (commented):
@adambaumeister the PR looks great! one minor comment.

bakatzir reviewed (commented):
Demo 20th April 2022 Notes:
Attendees: @GuyAfik @adambaumeister and myself

- the original issue of adding `target` arg to all cmds that have the `device_filter` argument only, to have a "closed option" for customers. for reference: https://github.com/demisto/etc/issues/36402
- `hostid` argument(in all cmds) - add examples in the arg desc “e.g, 12345678 or 192.168.0.1”
- `platform-reboot` command - add execution: true and update desc to say that this cmd is potentially disruptive.

adambaumeister commented:
Hi guys,

I added "target" as an argument for all operational commands to align them with the other commands in the pack.

I also changed the argument for reboot and update-ha-state to just be "target" instead of hostid.

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8611,14 +8634,331 @@ def resolve_container_name(container: Union[Panorama, Firewall, DeviceGroup, Tem
-    def get_device_groups(topology: Topology, device_filter_str: str = None) -> List[DeviceGroupInformation]:
+    def get_device_groups(
+            topology: Topology,
+            device_filter_str: str = None,
```
```suggestion
            device_filter_str: Optional[str] = None,
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8640,7 +8980,10 @@ def get_device_groups(topology: Topology, device_filter_str: str = None) -> List
-    def get_template_stacks(topology: Topology, device_filter_str: str = None) -> List[TemplateStackInformation]:
+    def get_template_stacks(
+            topology: Topology,
+            device_filter_str: str = None,
```
```suggestion
            device_filter_str: Optional[str] = None,
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -8776,15 +9119,20 @@ class UniversalCommand:
-    def get_system_info(topology: Topology, device_filter_str: str = None) -> ShowSystemInfoCommandResult:
+    def get_system_info(
+            topology: Topology,
+            device_filter_str: str = None,
```
```suggestion
            device_filter_str: Optional[str] = None,
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -6923,6 +6923,8 @@ script:
         name: device_filter_string
         required: false
         secret: false
+      - description: Target number of the firewall. Use only on a Panorama instance.
+        name: target
```
```suggestion
      - description: Target number of the firewall. Use only on a Panorama instance.
        name: target
        required: false
```
Add this to all below as well.

adambaumeister replied:
Should be done, but can you double check?

GuyAfik replied:
you did great, marking as resolved.

GuyAfik reviewed (commented):
@adambaumeister incredible work!

make sure you add the ```target``` argument to the README of the platform commands and to the existing commands if necessary 
Other than that looks great!

GuyAfik commented on Packs/PAN-OS/ReleaseNotes/1_12_0.md:
```diff
@@ -0,0 +1,10 @@
+
+#### Integrations
+##### Palo Alto Networks PAN-OS
+- Added 6 commands.
+  - ***pan-os-platform-install-software***
+  - ***pan-os-platform-reboot***
+  - ***pan-os-platform-get-system-status***
+  - ***pan-os-hygiene-check-log-forwarding***
+  - ***pan-os-hygiene-check-vulnerability-profiles***
+  - ***pan-os-platform-update-ha-state***
```
```suggestion
- Added 6 commands.
  - ***pan-os-platform-install-software***
  - ***pan-os-platform-reboot***
  - ***pan-os-platform-get-system-status***
  - ***pan-os-hygiene-check-log-forwarding***
  - ***pan-os-hygiene-check-vulnerability-profiles***
  - ***pan-os-platform-update-ha-state***
- Added the **target** arguments for the following commands:
   - command1
   - command2
```

GuyAfik commented on Packs/PAN-OS/ReleaseNotes/1_12_0.md:
```diff
@@ -0,0 +1,21 @@
+  - ***pan-os-hygiene-check-log-forwarding***
+  - ***pan-os-hygiene-check-vulnerability-profiles***
+  - ***pan-os-platform-update-ha-state***
+- Added *target* argument to the following commands
```
```suggestion
- Added the *target* argument to the following commands.
```

GuyAfik commented on Packs/PAN-OS/Integrations/Panorama/README.md:
```diff
@@ -5772,3 +5791,325 @@ Download the provided software version onto the device.
+| --- | --- | --- |
+| version | Software version to upgrade to, for example: 9.1.2. | Required | 
+| device_filter_string | SString by which to filter to only install to specific devices or  serial numbers. | Optional | 
+| sync | If provided, runs the download synchronously. Make sure 'execution-timeout' is increased. | Optional | 
```
missing documentation on the **target** argument here

adambaumeister replied:
good catch!

GuyAfik commented on Packs/PAN-OS/ReleaseNotes/1_12_0.md:
```diff
@@ -0,0 +1,21 @@
+  - ***pan-os-platform-get-available-software***
+  - ***pan-os-platform-get-ha-state***
+  - ***pan-os-platform-get-jobs***
+  - ***pan-os-platform-install-software***
```
if this is a command that you just added now (as part of the this PR's new commands), then you don't need to state it in the release-notes.

only those who you have added the **target** argument which were already implemented before (not in this PR)

GuyAfik reviewed (commented):
very few comments and we are done :)

make sure the **target** argument is also added to the READMEt of the new commands.

in addition please merge from master branch so the validations would run.

GuyAfik reviewed (approved):
simply incredible @adambaumeister 
Well done mate!
